modbus: add support for defining custom functions

lindblandro opened the pull request:
Enables support for custom function codes. Modbus specification allows vendor specific function codes in the range 65-72 & 100-110 [1] and this feature allows users to implement custom logic for those codes. Additionally, since the Zephyr Modbus stack doesn't implement all defined Modbus fcs this feature allows users to add support for codes outside the basic register reading / writing functionality offered by Zephyr.

Custom function codes can be added on a per-interface basis and the handler structures are allocated by the caller. The TCP server sample contains a simple usage example with a custom function that just increments a counter every time it's read.

[1]: https://modbus.org/docs/Modbus_Application_Protocol_V1_1b.pdf

Fixes #51319

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

lindblandro commented:
Related to https://github.com/zephyrproject-rtos/zephyr/issues/51319

lindblandro commented:
Anyone looking at this?

jfischer-no commented on samples/subsys/modbus/tcp_server/src/main.c:
```diff
@@ -164,7 +193,19 @@ static int init_modbus_server(void)
 	}
 
-	return modbus_init_server(server_iface, server_param);
+	int rc = modbus_init_server(server_iface, server_param);
+
+	if (rc < 0) {
+		return rc;
+	}
```
Please add `int err;`  at the top of this function,

```c
	err = modbus_init_server(server_iface, server_param);
	if (err) {
		return err;
	}
```

jfischer-no commented on subsys/modbus/Kconfig:
```diff
@@ -78,6 +78,13 @@ config MODBUS_FC08_DIAGNOSTIC
+config MODBUS_USER_DEFINED_FC
+	bool "User defined custom function code support"
+	depends on MODBUS_SERVER
+	default n
```
Redundant `default n`, I actually prefer to have this enabled by default, maybe even as part of core support without new Kconfig option.

lindblandro replied:
That sounds good to me. The Kconfig feature can always be introduced later.

jfischer-no commented on subsys/modbus/modbus_core.c:
```diff
@@ -214,7 +214,16 @@ static struct modbus_context *modbus_init_iface(const uint8_t iface)
+#ifdef CONFIG_MODBUS_USER_DEFINED_FC
+static int modbus_user_fc_init(struct modbus_context *ctx, struct modbus_iface_param param)
+{
+	LOG_DBG("Initializing user-defined function code support.");
+	sys_slist_init(&ctx->user_defined_cbs);
+	return 0;
```
```c
	sys_slist_init(&ctx->user_defined_cbs);
	LOG_DBG("Initializing user-defined function code support.");

	return 0;
```

jfischer-no commented on subsys/modbus/modbus_core.c:
```diff
@@ -282,6 +298,27 @@ int modbus_init_server(const int iface, struct modbus_iface_param param)
+{
+	struct modbus_context *ctx = modbus_get_context(iface);
+
+	if (!handler) {
```
Not sure it is necessary.

jfischer-no commented on subsys/modbus/modbus_core.c:
```diff
@@ -282,6 +298,27 @@ int modbus_init_server(const int iface, struct modbus_iface_param param)
+
+	LOG_DBG("Registered new custom function code %d", handler->function_code);
+	sys_slist_append(&ctx->user_defined_cbs, &handler->node);
+	return 0;
```
```c
	LOG_DBG("Registered new custom function code %d", handler->function_code);
	sys_slist_append(&ctx->user_defined_cbs, &handler->node);

	return 0;
```

jfischer-no commented on subsys/modbus/modbus_server.c:
```diff
@@ -926,6 +926,26 @@ static bool mbs_fc16_hregs_write(struct modbus_context *ctx)
 }
 
+#ifdef CONFIG_MODBUS_USER_DEFINED_FC
+static struct modbus_custom_fc *mbs_search_custom_fc(struct modbus_context *ctx, uint8_t fc)
```
```c
static struct modbus_custom_fc *mbs_try_user_fc(struct modbus_context *const ctx,
						const uint8_t fc)
```

jfischer-no commented on subsys/modbus/modbus_server.c:
```diff
@@ -995,10 +1016,24 @@ bool modbus_server_handler(struct modbus_context *ctx)
-		mbs_exception_rsp(ctx, MODBUS_EXC_ILLEGAL_FC);
-		send_reply = true;
-		break;
+		if (IS_ENABLED(CONFIG_MODBUS_USER_DEFINED_FC)) {
```
We should just call `send_reply = mbs_try_user_fc(...);` and handle user callback as well mbs_exception_rsp() there.

jfischer-no commented on subsys/modbus/modbus_server.c:
```diff
@@ -995,10 +1016,24 @@ bool modbus_server_handler(struct modbus_context *ctx)
+				send_reply = handler->callback(iface, &ctx->rx_adu, &ctx->tx_adu,
+							       handler->user_data);
+			} else {
+				LOG_DBG("Didn't find any handlers for code %u", fc);
```
Exception not called.

vai-helin committed:
modbus: add support for defining custom functions

Enables support for custom function codes. Modbus specification allows
vendor specific function codes in the range 65-72 & 100-110 [1] and this
feature allows users to implement custom logic for those codes.
Additionally, since the Zephyr Modbus stack doesn't implement all defined
Modbus fcs this feature allows users to add support for codes outside the
basic register reading / writing functionality offered by Zephyr.

Custom function codes can be added on a per-interface basis and the handler
structures are allocated by the caller.

[1]: https://modbus.org/docs/Modbus_Application_Protocol_V1_1b.pdf

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

lindblandro commented on samples/subsys/modbus/tcp_server/src/main.c:
```diff
@@ -26,6 +26,35 @@ static const struct gpio_dt_spec led_dev[] = {
+		return false;
+	}
+
+	uint8_t subfunc = rx_adu->data[0];
+	uint8_t data_len = rx_adu->data[1];
```
I'll place them there, but is there a particular reason for this? The values are only logged directly after this and generally speaking (IMO, at least) definitions as close to usage yields more readable code. Randomly sampling some code would suggest that declaring variables at the top is the norm, but I couldn't find anything on this in the coding style section https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html#coding-style

lindblandro commented on samples/subsys/modbus/tcp_server/src/main.c:
```diff
@@ -146,7 +175,7 @@ const static struct modbus_iface_param server_param = {
 	.server = {
 		.user_cb = &mbs_cbs,
-		.unit_id = 1,
+		.unit_id = 255,
```
I think I made this change to satisfy pymodbus which initialized it to 255 and just forgot it was there. The Zephyr Modbus stack didn't seem to respect the protocol as it states that

>Unit Identifier – This field is used for intra-system routing purpose. It is typically
>used to communicate to a MODBUS+ or a MODBUS serial line slave through a
>gateway between an Ethernet TCP-IP network and a MODBUS serial line. This field is
>set by the MODBUS Client in the request and must be returned with the same value in
>the response by the server.

and Zephyr filtered responses based on the unit id value. Not sure if this is by design or if my test code was crappy so wouldn't exactly categorize it as a flaw just yet.

lindblandro commented on include/zephyr/modbus/modbus.h:
```diff
@@ -375,6 +375,11 @@ int modbus_iface_get_by_name(const char *iface_name);
 /**
  * @brief ADU raw callback function signature
  *
+ * This is used for separating the transmission channel from query
+ * handling. It is not intended for implementing vendor specific function
+ * codes, which is also subtly implied from the signature: it only receives
+ * the outgoing ADU as a const pointer.
+ *
```
Is it ok if I include them in a separate commit in this PR? I added the docstring to describe the difference between the two functionalities mostly to clarify the implementation for myself, but it could be useful to others as well.

lindblandro replied:
I'll probably reword / improve the description a little bit also.

lindblandro replied:
Removed this.

vai-helin committed:
samples: modbus: add user function code example

Demonstrate user function codes with a simple example with a user defined
fc that maps to a function that just increments a counter every time it's
read.

Signed-off-by: Henrik Lindblom <henrik.lindblom@vaisala.com>

lindblandro commented on include/zephyr/modbus/modbus.h:
```diff
@@ -535,6 +586,24 @@ void modbus_raw_set_server_failure(struct modbus_adu *adu);
+ *
+ * @retval           0 on success
+ */
+int modbus_define_function_code(const int iface, struct modbus_custom_fc *handler);
```
Better indeed, How about plain `modbus_register_fc()` or (to better signal that "register" is a verb and not a noun) `modbus_register_new_fc()`?

lindblandro commented:
@jfischer-no I made the requested changes, but did not mark the issues resolved as per contribution guidelines (except for the couple first ones which I resolved before finding the mention in the docs).

pdgendt commented on include/zephyr/modbus/modbus.h:
```diff
@@ -384,6 +384,48 @@ int modbus_iface_get_by_name(const char *iface_name);
+ *
+ * @retval           1 If response should be sent, 0 otherwise
+ */
+typedef int (*modbus_user_cb_t)(const int iface,
+				  const struct modbus_adu *rx_adu,
+				  struct modbus_adu *tx_adu,
+				  void *const user_data);
```
Nit: Fix indentation

lindblandro replied:
Fixed.

pdgendt replied:
Please update the commits as per the [contribution guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html). Do not add fixup commits.

I think the PR should contain only 2 commits: 
- Adding the user functions
- Update sample

jfischer-no commented on include/zephyr/modbus/modbus.h:
```diff
@@ -384,6 +384,48 @@ int modbus_iface_get_by_name(const char *iface_name);
+/**
+ * @brief Helper macro for initializing custom function code structs
+ */
+#define MODBUS_CUSTOM_FC(cb, fc, userdata)				\
```
Let's force it to be static and shorter for the user.

```c
#define MODBUS_CUSTOM_FC_DEFINE(name, user_cb, user_fc, user_data)		\
	static struct modbus_custom_fc modbus_cfc_##name = {			\
		.callback = user_cb,						\
		.user_data = user_data,						\
		.function_code = user_fc					\
	}
```

lindblandro commented on include/zephyr/modbus/modbus.h:
```diff
@@ -384,6 +384,48 @@ int modbus_iface_get_by_name(const char *iface_name);
+typedef int (*modbus_user_cb_t)(const int iface,
+				const struct modbus_adu *rx_adu,
+				struct modbus_adu *tx_adu,
+				void *const user_data);
```
Please check that I understood your idea correctly. The `excep_code` is now reset before every call to the callback function and I feel it might lead to surprises unless locks are added.

DaAwesomeP commented:
Just checking in on the status of this

DaAwesomeP commented:
Hey @lindblandro, just checking in again.

jfischer-no commented on subsys/modbus/modbus_server.c:
```diff
@@ -926,6 +926,36 @@ static bool mbs_fc16_hregs_write(struct modbus_context *ctx)
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&ctx->user_defined_cbs, p, node) {
+		if (p->fc == fc) {
+			LOG_DBG("Found custom handler");
+			int iface = modbus_iface_get_by_ctx(ctx);
+
+			p->excep_code = MODBUS_EXC_NONE;
+			bool rval = p->cb(iface, &ctx->rx_adu, &ctx->tx_adu, &p->excep_code,
+					p->user_data);
+
+			if (p->excep_code != MODBUS_EXC_NONE) {
```
I have pointed this out a few times, please move the variable definition to the top of a block.
```c
	if (p->fc == fc) {
			int iface = modbus_iface_get_by_ctx(ctx);
			bool rval;

			LOG_DBG("Found custom handler");
			p->excep_code = MODBUS_EXC_NONE;
			rval = p->cb(iface, &ctx->rx_adu, &ctx->tx_adu, &p->excep_code,
					p->user_data);
```

jfischer-no commented on samples/subsys/modbus/tcp_server/src/main.c:
```diff
@@ -26,6 +26,43 @@ static const struct gpio_dt_spec led_dev[] = {
 
+static int custom_read_count;
+
+static bool custom_handler(const int iface,
+			  const struct modbus_adu *rx_adu,
+			  struct modbus_adu *tx_adu,
+			  uint8_t *const excep_code,
+			  void *const user_data)
```
Please add one more space after tab to indent properly.

lindblandro replied:
Had my git diffs showing the wrong tab sizes, which caused formatting to be a bit hard to judge. Should be fixed now.

jfischer-no commented on samples/subsys/modbus/tcp_server/src/main.c:
```diff
@@ -164,7 +202,16 @@ static int init_modbus_server(void)
+
+	err = modbus_register_user_fc(server_iface, &modbus_cfg_custom);
+
+	/* ignore if disabled */
+	return (err == -ENOTSUP) ? 0 : err;
```
It look like left over from previous patch series, I guess it can be just
`return  modbus_register_user_fc(server_iface, &modbus_cfg_custom);`

lindblandro replied:
Ack.

jfischer-no reviewed (commented):
LGTM, just two minor comments.

kartben commented on include/zephyr/modbus/modbus.h:
```diff
@@ -384,6 +395,55 @@ int modbus_iface_get_by_name(const char *iface_name);
+ * @brief Custom function code definition.
+ */
+struct modbus_custom_fc {
+	sys_snode_t node;
+	modbus_custom_cb_t cb;
+	void *user_data;
+	uint8_t fc;
+	uint8_t excep_code;
```
Please document these or have them in a `@cond INTERNAL_HIDDEN` section if they are not meant to be API

jfischer-no replied:
@lindblandro Please guard it "in a @cond INTERNAL_HIDDEN section".

lindblandro replied:
Added the guard.

kartben commented on include/zephyr/modbus/modbus.h:
```diff
@@ -40,6 +40,17 @@ extern "C" {
 #define MODBUS_MBAP_LENGTH		7
 /** Length of MBAP Header plus function code */
 #define MODBUS_MBAP_AND_FC_LENGTH	(MODBUS_MBAP_LENGTH + 1)
+/* Modbus exception codes */
+#define MODBUS_EXC_NONE				0
+#define MODBUS_EXC_ILLEGAL_FC			1
+#define MODBUS_EXC_ILLEGAL_DATA_ADDR		2
+#define MODBUS_EXC_ILLEGAL_DATA_VAL		3
+#define MODBUS_EXC_SERVER_DEVICE_FAILURE	4
+#define MODBUS_EXC_ACK				5
+#define MODBUS_EXC_SERVER_DEVICE_BUSY		6
+#define MODBUS_EXC_MEM_PARITY_ERROR		8
+#define MODBUS_EXC_GW_PATH_UNAVAILABLE		10
+#define MODBUS_EXC_GW_TARGET_FAILED_TO_RESP	11
```
Please group these with `@name` + add doxygen docstrings to each define. See below for `@name` usage.

```suggestion
/** @name Modbus exception codes 
 * @{
*/
#define MODBUS_EXC_NONE				0
#define MODBUS_EXC_ILLEGAL_FC			1
#define MODBUS_EXC_ILLEGAL_DATA_ADDR		2
#define MODBUS_EXC_ILLEGAL_DATA_VAL		3
#define MODBUS_EXC_SERVER_DEVICE_FAILURE	4
#define MODBUS_EXC_ACK				5
#define MODBUS_EXC_SERVER_DEVICE_BUSY		6
#define MODBUS_EXC_MEM_PARITY_ERROR		8
#define MODBUS_EXC_GW_PATH_UNAVAILABLE		10
#define MODBUS_EXC_GW_TARGET_FAILED_TO_RESP	11
/** @} */
```

jfischer-no replied:
@kartben Documentation fixes may also be applied over the next two weeks.

lindblandro replied:
Added the section name.

jfischer-no reviewed (dismissed):
@lindblandro Thank you for your patience.

kartben reviewed (approved):
+1 for doc updates - thanks!

jfischer-no reviewed (approved):
Thanks
